lib/channel/src/lib.rs

```diff
@@ -508,40 +527,59 @@ pub fn channel<T: Serialize>(limit: Limit, unique: bool) -> (Sender<T>, Receiver
 // to `recv` but the queue is empty
 pub struct OnDemandReceiver<T: Serialize> {
     channel: Arc<Channel<T>>,
-    listener: EventListener,
+    listener: Option<EventListener>,
 }
 
 impl<T: Serialize> Clone for OnDemandReceiver<T> {
     fn clone(&self) -> Self {
+        self.channel.increment_on_demand_count();
         Self {
             channel: self.channel.clone(),
-            listener: self.channel.on_demand_listen(),
+            listener: None,
         }
     }
 }
 
+// whenever a `OnDemandReceiver` is dropped, be sure to decrement the on_demand sender count
+// an OnDemandReceiver is also a Sender, so Drop for Sender will notify `Receiver`s that are waiting for data.
+impl<T: Serialize> Drop for OnDemandReceiver<T> {
+  fn drop(&mut self) {
+    self.channel.decrement_on_demand_count();
+  }
+}
+
 impl<T: Serialize + Send + 'static> Stream for OnDemandReceiver<T> {
     type Item = ();
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+      loop {
         // end the stream if there are no more receivers
         if self.channel.receiver_count() == 0 {
+            self.listener = None;
             return Poll::Ready(None);
         }
 
-        let ret = Pin::new(&mut self.listener).poll(cx).map(Some);
-        if ret.is_ready() {
-            self.listener = self.channel.on_demand_listen();
-        };
-        ret
+        if let Some(listener) = self.listener.as_mut() {
+          let ret = Pin::new(listener).poll(cx).map(Some);
+          if ret.is_ready() {
+            self.listener = Some(self.channel.on_demand_listen());
+          };
+          return ret;
+        } else if self.channel.receiver_count() == 0 {
```

Comment:
Does this need to be checked here since it's also checked on L557

Reply:
I was bringing over the code from the normal receiver poll and combining it with the code from this poll. I saw it had both, but remembered a section of code where you check the receiver count twice because the receiver could've gone away since then. But we have the loop so it's probably not needed.